x-pack/libbeat/common/aws/acker.go

```diff
@@ -18,54 +18,40 @@ import (
 // more S3 objects.
 type EventACKTracker struct {
 	sync.Mutex
-	PendingACKs int64
-	ctx         context.Context
-	cancel      context.CancelFunc
+	PendingACKs *atomic.Uint64
 }
 
-func NewEventACKTracker(ctx context.Context) *EventACKTracker {
-	ctx, cancel := context.WithCancel(ctx)
-	return &EventACKTracker{ctx: ctx, cancel: cancel}
+func NewEventACKTracker() *EventACKTracker {
+	return &EventACKTracker{PendingACKs: atomic.NewUint64(0)}
 }
 
 // Add increments the number of pending ACKs.
 func (a *EventACKTracker) Add() {
-	a.Lock()
-	a.PendingACKs++
-	a.Unlock()
+	a.PendingACKs.Inc()
 }
 
 // ACK decrements the number of pending ACKs.
 func (a *EventACKTracker) ACK() {
-	a.Lock()
-	defer a.Unlock()
-
-	if a.PendingACKs <= 0 {
+	if a.PendingACKs.Load() <= 0 {
 		panic("misuse detected: negative ACK counter")
 	}
 
-	a.PendingACKs--
-	if a.PendingACKs == 0 {
-		a.cancel()
-	}
+	a.PendingACKs.Dec()
 }
 
 // Wait waits for the number of pending ACKs to be zero.
-// Wait must be called sequentially only after every expected
+// Wait must be called only after every expected
 // `Add` calls are made. Failing to do so could reset the pendingACKs
 // property to 0 and would results in Wait returning after additional
 // calls to `Add` are made without a corresponding `ACK` call.
 func (a *EventACKTracker) Wait() {
-	// If there were never any pending ACKs then cancel the context. (This can
-	// happen when a document contains no events or cannot be read due to an error).
-	a.Lock()
-	if a.PendingACKs == 0 {
-		a.cancel()
-	}
-	a.Unlock()
 
-	// Wait.
-	<-a.ctx.Done()
+	// If there were no any pending ACKs returns.
+	for {
+		if a.PendingACKs.Load() == 0 {
+			return
+		}
+	}
 }
```

Comment:
This is just busy waiting, have you observed the CPU utilization of the input spiking when there is a goroutine blocked on `Wait`? That is what I would expect to happen.

Getting rid of the mutex is a good idea but this part of the implementation needs to be improved.

Reply:
Per my comment in https://github.com/elastic/beats/pull/38146#discussion_r1511815527 could you check for the number of pending acks to reach zero each time an acknowledgement is received instead of busy waiting in a separate goroutine?

Reply:
You can store information identifying the originating s3 object so that you know what is being acked if necessary.

The filestream acker is a complex reference example in that is receiving acks from multiple file sources and then checkpointing the progress to disk in a separate goroutine.

https://github.com/elastic/beats/blob/4f46fcb888b7f12706d005b7e1d2d1dafcbe6fcd/filebeat/input/filestream/internal/input-logfile/input.go#L97-L122

Reply:
> This is just busy waiting, have you observed the CPU utilization of the input spiking when there is a goroutine blocked on `Wait`? That is what I would expect to happen.
> 
> Getting rid of the mutex is a good idea but this part of the implementation needs to be improved.

yes, CPU utilization spikes. let me test the specific sqs acker